File: src/lib/models/Discipline.ts
import BaseEntity from "./BaseEntity";

export default class Discipline extends BaseEntity {
    public readonly userId: number;
    public readonly name: string;
    public readonly description: string | null;

    private constructor(
        id: number,
        userId: number,
        name: string,
        description: string | null,
        createdAt: Date,
        updatedAt: Date
    ) {
        super(id, createdAt, updatedAt);
        this.userId = userId;
        this.name = name;
        this.description = description;
    }

    static create(
        id: number,
        userId: number,
        name: string,
        description: string | null,
        createdAt: Date,
        updatedAt: Date
    ): Discipline {
        return new Discipline(
            id,
            userId,
            name,
            description,
            createdAt,
            updatedAt
        );
    }
}

File: src-tauri/src/model/discipline.rs
use chrono::{DateTime, Utc};

#[derive(Debug, Clone, sqlx::FromRow)]
pub struct Discipline{
    pub id: i64,
    pub user_id: i64,
    pub name: String,
    pub description: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub total_items: i32,
    pub items_mastered: i32,
    pub progress_percent: f64,
    pub last_review_date: Option<String>,
}

#[derive(Debug, Clone, sqlx::FromRow)]
pub struct UpdateDiscipline {
    pub name: Option<String>,
    pub description: Option<String>,
}

impl UpdateDiscipline {
    pub fn new(name: Option<String>, description: Option<String>,) -> Self {
        Self { name, description }
    }
}